{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Editing efficiency analysis\n",
    "\n",
    "This notebook contains the code to analyze the overall editing efficiency (by C position) for all the validation sgRNAs that we tested. This notebook takes as input the nucleotide_percentage files generated by the BEV_nucleotide_percentage_plots notebook."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from os import path\n",
    "from pathlib import Path\n",
    "\n",
    "from IPython.display import Javascript, display\n",
    "\n",
    "mpl.rc('pdf', fonttype=42)\n",
    "mpl.rcParams['font.sans-serif'] = \"Arial\"\n",
    "mpl.rcParams['font.family'] = \"sans-serif\"\n",
    "sns.set_context('paper')\n",
    "sns.set_style('ticks')\n",
    "boxprops = {'edgecolor': 'k', 'linewidth': 0.5, 'facecolor': 'w'}\n",
    "lineprops = {'color': 'k', 'linewidth': 0.5}\n",
    "stripplot_kwargs = dict({'linewidth': 0.5, 'size': 3, 'alpha': 0.8})\n",
    "boxplot_kwargs = dict({'boxprops': {'linewidth':0,'facecolor':'w'}, 'medianprops': {'linewidth':1},\n",
    "                       'whiskerprops': {'linewidth':0}, 'capprops': {'linewidth':0},\n",
    "                       'width': 0.8,'whis':(10,90)})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <script>\n",
       "            function code_toggle_2752437172129763717() {\n",
       "                $('div.cell.code_cell.rendered.selected').find('div.input').toggle();\n",
       "            }\n",
       "\n",
       "            \n",
       "        </script>\n",
       "\n",
       "        <a href=\"javascript:code_toggle_2752437172129763717()\">Show/Hide Toggle Function</a>\n",
       "    "
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from IPython.display import HTML\n",
    "import random\n",
    "\n",
    "def hide_toggle(toggle_text_addon = '', for_next=False):\n",
    "    this_cell = \"\"\"$('div.cell.code_cell.rendered.selected')\"\"\"\n",
    "    next_cell = this_cell + '.next()'\n",
    "    \n",
    "    toggle_text = 'Show/Hide' + ' ' + toggle_text_addon  # text shown on toggle link\n",
    "    target_cell = this_cell  # target cell to control with toggle\n",
    "    js_hide_current = ''  # bit of JS to permanently hide code in current cell (only when toggling next cell)\n",
    "\n",
    "    if for_next:\n",
    "        target_cell = next_cell\n",
    "        toggle_text += ' next cell'\n",
    "        js_hide_current = this_cell + '.find(\"div.input\").hide();'\n",
    "\n",
    "    js_f_name = 'code_toggle_{}'.format(str(random.randint(1,2**64)))\n",
    "\n",
    "    html = \"\"\"\n",
    "        <script>\n",
    "            function {f_name}() {{\n",
    "                {cell_selector}.find('div.input').toggle();\n",
    "            }}\n",
    "\n",
    "            {js_hide_current}\n",
    "        </script>\n",
    "\n",
    "        <a href=\"javascript:{f_name}()\">{toggle_text}</a>\n",
    "    \"\"\".format(\n",
    "        f_name=js_f_name,\n",
    "        cell_selector=target_cell,\n",
    "        js_hide_current=js_hide_current, \n",
    "        toggle_text=toggle_text\n",
    "    )\n",
    "\n",
    "    return HTML(html)\n",
    "hide_toggle(toggle_text_addon='Toggle Function')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <script>\n",
       "            function code_toggle_14457475437358309053() {\n",
       "                $('div.cell.code_cell.rendered.selected').find('div.input').toggle();\n",
       "            }\n",
       "\n",
       "            \n",
       "        </script>\n",
       "\n",
       "        <a href=\"javascript:code_toggle_14457475437358309053()\">Show/Hide check_folder_filepath function</a>\n",
       "    "
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check to make sure folder filepaths end in '/' but do not begin with a '/' for relative path\n",
    "def check_folder_filepath(filepath):\n",
    "    # Check if filepath is relative\n",
    "    if not (path.isabs(filepath)):\n",
    "        # if relative, get rid of '/' at the beginning of filepath\n",
    "        if filepath[0] == '/':\n",
    "            filepath = filepath[1:]\n",
    "    if filepath[-1] != '/':\n",
    "        filepath = filepath+'/'\n",
    "    # Check if filepath exists\n",
    "    if path.exists(filepath):\n",
    "        return filepath\n",
    "    else:\n",
    "        raise ValueError('Filepath does not exist. Please check your inputs.')\n",
    "        return\n",
    "\n",
    "\n",
    "hide_toggle(toggle_text_addon='check_folder_filepath function')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## User inputs\n",
    "\n",
    "<font color='blue'> Please follow steps indicated in blue, then run the notebook to generate output files. If the files are formatted as described in the documentation, the code in the 'Functions' section should not need to be altered. </font> "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Metainformation file** \n",
    "\n",
    "<font color='blue'> <b>Step 1:</b> Please enter the filepath to the metainformation input file used in the allele frequencies notebook. The same file will be used in this notebook. </font> "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Please enter input filepath here: ../../AudreyData/TP53/Metainfo_input_ABE_TP53_fixed_updated_1d_sample.csv\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sg</th>\n",
       "      <th>sgRNA_sequence</th>\n",
       "      <th>translation_ref_seq</th>\n",
       "      <th>BEV_start</th>\n",
       "      <th>BEV_end</th>\n",
       "      <th>primer</th>\n",
       "      <th>frame</th>\n",
       "      <th>first_codon</th>\n",
       "      <th>last_codon</th>\n",
       "      <th>rev_com</th>\n",
       "      <th>BEV_ref</th>\n",
       "      <th>BEV_test</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1d</td>\n",
       "      <td>GCTCCTCCATGGCAGTGACC</td>\n",
       "      <td>[TTCCTCTTGCAGCAGCCAGACTGCCTTCCGGGTCACTGCC]ATGG...</td>\n",
       "      <td>417</td>\n",
       "      <td>426</td>\n",
       "      <td>F3_R2</td>\n",
       "      <td>1</td>\n",
       "      <td>ATG</td>\n",
       "      <td>CTG</td>\n",
       "      <td>True</td>\n",
       "      <td>417;418</td>\n",
       "      <td>425;426</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   sg        sgRNA_sequence  \\\n",
       "0  1d  GCTCCTCCATGGCAGTGACC   \n",
       "\n",
       "                                 translation_ref_seq  BEV_start  BEV_end  \\\n",
       "0  [TTCCTCTTGCAGCAGCCAGACTGCCTTCCGGGTCACTGCC]ATGG...        417      426   \n",
       "\n",
       "  primer  frame first_codon last_codon  rev_com  BEV_ref BEV_test  \n",
       "0  F3_R2      1         ATG        CTG     True  417;418  425;426  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "global old_input_file\n",
    "# input_filepath = 'AudreyData/Metainfo_input_f2_r1_ABE.csv'\n",
    "input_filepath = input(\"Please enter input filepath here: \")\n",
    "old_input_file = pd.read_csv(input_filepath)\n",
    "\n",
    "old_input_file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<font color='blue'><b>Step 2:</b> Enter filepath to folder where the files generated by the nucleotide percentage notebook were stored. Please make sure that the filepath ends in a '/'. </font> "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Please enter filepath to folder with nucleotide percentage outputs here: ../../AudreyData/TP53/ABE/nucleotide_percentage\n",
      "../../AudreyData/TP53/ABE/nucleotide_percentage/\n"
     ]
    }
   ],
   "source": [
    "global nuc_per_filepath\n",
    "# nuc_per_filepath = 'AudreyData/Validation_CRISPResso_results/nucleotide_percentage/'\n",
    "nuc_per_filepath = input(\"Please enter filepath to folder with nucleotide percentage outputs here: \")\n",
    "nuc_per_filepath = check_folder_filepath(nuc_per_filepath)\n",
    "print(nuc_per_filepath)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<font color='blue'><b>Step 3:</b> Enter filepath to folder where the files generated by this notebook will be stored. Please make sure that the filepath ends in a '/'. If the folders in this file path do not currently exist, they will be created when the notebook is run.  </font> "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Please enter output folder filepath here: ../../AudreyData/TP53/ABE/\n",
      "../../AudreyData/TP53/ABE/\n"
     ]
    }
   ],
   "source": [
    "global output_filepath\n",
    "# output_filepath = 'AudreyData/Validation_CRISPResso_results/'\n",
    "output_filepath = input(\"Please enter output folder filepath here: \")\n",
    "output_filepath = check_folder_filepath(output_filepath)\n",
    "print(output_filepath)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<font color='blue'><b>Step 4:</b> Enter the name of the gene targeted in experiment. </font> "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Please enter gene name here: TP53\n"
     ]
    }
   ],
   "source": [
    "global experiment\n",
    "experiment = input(\"Please enter gene name here: \")\n",
    "# next_gene = input(\"Are there more genes targeted in this experiment? Please enter y for 'yes' or n for 'no.'\")\n",
    "# if next_gene == 'y':\n",
    "    \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<font color='blue'><b>Step 5:</b> Please select the type of base editor (BE) used in the samples in input file. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Please specify the type of base editor used in the input samples by entering 'A' for A base editor or 'C' for C base editor: A\n"
     ]
    }
   ],
   "source": [
    "global be_type\n",
    "\n",
    "be_type_input = input(\"Please specify the type of base editor used in the input samples by entering 'A' for A base editor or 'C' for C base editor: \")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/javascript": [
       "IPython.notebook.execute_cells_below()"
      ],
      "text/plain": [
       "<IPython.core.display.Javascript object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Make sure a base editor is selected and not default value\n",
    "\n",
    "if (be_type_input != 'A') and (be_type_input != 'C'):\n",
    "    raise Exception('Invalid input. Please enter either A or C to specify base editor.')\n",
    "\n",
    "else:\n",
    "    # Run rest of notebook! \n",
    "    be_type = be_type_input + 'BE'\n",
    "    display(Javascript('IPython.notebook.execute_cells_below()'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "This function removes any NaN rows from input_file\n",
    "'''\n",
    "def clean_input_file(df):\n",
    "    df = df.dropna() #drop NaN \n",
    "    #dropna() converts int to float, so convert them back\n",
    "    float_cols = df.select_dtypes(include=['float64']).columns #select subset of df of type float\n",
    "    for col in float_cols: \n",
    "        #get original column index so can replace at correct loc\n",
    "        index = df.columns.get_loc(col)\n",
    "        #rename float cols as \"float_\"col name \n",
    "        float_col_name = 'float_' + col\n",
    "        df = df.rename(columns = {col : float_col_name})\n",
    "        #overwrite as type int\n",
    "        float_to_int = df[float_col_name].astype(int).copy()\n",
    "        df.insert(index, col, float_to_int)\n",
    "        #drop float column \n",
    "        df = df.drop(float_col_name, axis = 1)\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "def run(old_input_file):\n",
    "    \n",
    "    #check for NaN values i.e. blank rows\n",
    "    if old_input_file.isnull().values.any(): \n",
    "        old_input_file = clean_input_file(old_input_file)\n",
    "\n",
    "    input_file = old_input_file.copy()\n",
    "    \n",
    "    #changed bev_test to bev_ref\n",
    "    input_file['BEV_reps'] = input_file['BEV_ref']\n",
    "    input_file['Experiment'] = experiment\n",
    "    offset = []\n",
    "    nuc_list = []\n",
    "    sg_list = []\n",
    "    sgrna_seq_list=[]\n",
    "    edits_list = []\n",
    "    expt_list = []\n",
    "\n",
    "    for i,r in input_file.iterrows():\n",
    "        # Open corresponding \"nucleotide_percentage\" file\n",
    "        #file_path = '../Data/Validation_CRISPResso_results/nucleotide_percentage/BEV_'+'_'.join(r['BEV_reps'].split(';'))+'_'+r['primer']+'.csv'\n",
    "        file_path = nuc_per_filepath+'BEV_'+'_'.join(r['BEV_reps'].split(';'))+'_'+r['primer']+'.csv'\n",
    "        print(file_path)\n",
    "        data = pd.read_csv(file_path)\n",
    "        \n",
    "        # Filter for positions where WT nuc is A or C depending on base editor type \n",
    "        if be_type == 'ABE':\n",
    "            data = data[data['WT'] == 'A']\n",
    "        elif be_type == 'CBE':\n",
    "            data = data[data['WT'] == 'C']\n",
    "\n",
    "        for nuc in ['A','G','T','N','del']:\n",
    "            col = nuc+'_avg'\n",
    "            edits_list.extend(data[col].tolist())\n",
    "            offset.extend(data['offset_position'].tolist())\n",
    "            nuc_list.extend([nuc] * len(data))\n",
    "            sg_list.extend([r['sg']] * len(data))\n",
    "            sgrna_seq_list.extend([r['sgRNA_sequence']] * len(data))\n",
    "            expt_list.extend([r['Experiment']] * len(data))\n",
    "    # Scale to percentage\n",
    "    edits_list = [x*100 for x in edits_list]\n",
    "    tuples_list = list(zip(edits_list,offset,nuc_list,sg_list,sgrna_seq_list,expt_list))\n",
    "#     tuples_list = list(zip(edits_list,offset,nuc_list, sgrna_seq_list,expt_list))\n",
    "    df = pd.DataFrame(tuples_list, columns = ['Percentage', 'Position','Nucleotide','sg','sgRNA_sequence','Experiment'])\n",
    "#     df = pd.DataFrame(tuples_list, columns = ['Percentage', 'Position','Nucleotide','sgRNA_sequence','Experiment']).drop_duplicates()\n",
    "    \n",
    "    if be_type == 'ABE':\n",
    "        df[df['Nucleotide'] == 'G'].sort_values(by='Percentage',ascending=False).head()\n",
    "    elif be_type == 'CBE':\n",
    "        df[df['Nucleotide'] == 'T'].sort_values(by='Percentage',ascending=False).head()\n",
    "    \n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "../../AudreyData/TP53/ABE/nucleotide_percentage/BEV_417_418_F3_R2.csv\n"
     ]
    }
   ],
   "source": [
    "df = run(old_input_file)\n",
    "save_filepath = output_filepath + 'editing_efficiency_input.csv'\n",
    "df.to_csv(save_filepath)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "df = df.copy().drop('sg', axis=1).drop_duplicates()\n",
    "if be_type == 'ABE':\n",
    "    df.loc[(df['Position'] >= 4) & (df['Position'] <= 8) & (df['Nucleotide'] == 'G'),'Position'].value_counts()\n",
    "elif be_type == 'CBE':\n",
    "    df.loc[(df['Position'] >= 4) & (df['Position'] <= 8) & (df['Nucleotide'] == 'T'),'Position'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Generate plots "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_plot(nuc,experiments):\n",
    "    color_dict = {'A':sns.color_palette('Set2')[0],\n",
    "                  'G':sns.color_palette('Set2')[1],\n",
    "                  'T':sns.color_palette('Set2')[2],\n",
    "                  'del':sns.color_palette('Set2')[4]}\n",
    "    fig,ax=plt.subplots(figsize=(4,1.5))\n",
    "    sns.boxplot(x='Position',y='Percentage',data=df.loc[(df['Position'] >= -10) & (df['Position'] <= 20) &\n",
    "                                                        (df['Nucleotide'] == nuc) & (df['Experiment'].isin(experiments))],\n",
    "                fliersize=0,**boxplot_kwargs)\n",
    "    sns.stripplot(x='Position',y='Percentage',data=df.loc[(df['Position'] >= -10) & (df['Position'] <= 20) &\n",
    "                                                          (df['Nucleotide'] == nuc) & (df['Experiment'].isin(experiments))],\n",
    "                  color=color_dict[nuc],**stripplot_kwargs)\n",
    "    ax.set_ylim(0,100)\n",
    "    if nuc == 'del':\n",
    "        ax.set_title('deletions',fontsize=8)\n",
    "    else:\n",
    "        if be_type == 'ABE':\n",
    "            ax.set_title('A>'+nuc,fontsize=8)\n",
    "        elif be_type == 'CBE':\n",
    "            ax.set_title('C>'+nuc,fontsize=8)\n",
    "    ax.set_ylabel('% conversion',fontsize=6)\n",
    "    \n",
    "    if be_type == 'ABE':\n",
    "        ax.set_xlabel('position of A', fontsize=6)\n",
    "    elif be_type == 'CBE':\n",
    "        ax.set_xlabel('position of C', fontsize=6)\n",
    "    plt.xticks(fontsize=6)\n",
    "    plt.yticks(fontsize=6)\n",
    "    plt.tight_layout()\n",
    "    # format filepath so it's saved as 'folder/BRCA1_T_editing_eff.pdf'\n",
    "    filepath = output_filepath+'_'.join(experiments)+'_'+nuc+'_editing_eff.pdf'\n",
    "    Path(output_filepath).mkdir(parents=True, exist_ok=True)\n",
    "    fig.savefig(filepath, transparent = True)\n",
    "    return fig,ax"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<font color='blue'><b>Step 6:</b> Enter the target nucleotide of interest (T, A, G, C, or del). You can re-run the following cells to change the nucleotide being plotted. </font> "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Please enter target nucleotide here: G\n"
     ]
    }
   ],
   "source": [
    "nuc = input(\"Please enter target nucleotide here: \")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAARkAAABlCAYAAAB5h5cZAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjQuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8rg+JYAAAACXBIWXMAAAsTAAALEwEAmpwYAAAKxElEQVR4nO3dXVBUZQMH8P8BOSAQH4u4SMNMjILrWNiQhpmvMeakiCMjkB/kuFhtZeNFU03clu9NWtZWNKWOugzOMDbjUEZcMA7QEOMI4ih9ATLStMiHbLhYK+zZZfe9aKCatz0usucc9vD/3Tgi+5z/0XP+PPA8xxX8fr8fREQKidA6ABHpG0uGiBTFkiEiRbFkiEhRLBkiUhRLhogUtUDrAKQvTqcT27Ztg9VqRUZGBm7fvg2TyTT9536/H1arFYODgxBFEW63GxUVFVi0aJGGqUlJAvfJUCgdO3YMycnJaG5uxpEjR1BTU4Pr169jzZo1KCwsRH19PSRJQllZGQDAbrfD5XL9o4hIXziToZDxeDxoaGjAF198gcbGRgwMDMBisaC3txfvvPMO3G43enp6YLFYAACVlZX45ZdfkJGRwZLRMc5kKGTOnz+P06dP45FHHsHg4CASEhIQERGBrKwslJSUICUlBWfPnsXExATMZjMAoL+/H5WVlXj33Xc1Tk9KYclQyOzduxcffvghUlNT4ff7sXv3brz//vvIyMiY/hyfzwer1YobN24gOjoaHo8HL774InJycjRMTkpiyRCRoriETUSKYskQkaJYMkSkqJAsYUuSBKvVitzcXPzwww9wu92wWCyor6+H1+uF0WhEQUFBKA5FRGEmJDMZh8MBk8kEl8uF/Px8FBcXo7m5GTdv3kR5eTk6OztDcRgiCkMhmcmkp6fDaDRicHAQERF/9VZkZGTA13zyySeorKwMxeGJaA7o7u7+14+HbAn70qVLGBoaQldXF0RRhNlsRm1tLURRREpKCrZu3XrPMZYvXx4wKBHNXXL37pzaJ8OSIQpPcvcuV5eISFEsGSJSFEuGiBTFkiEiRbFkiEhRLBkiUhRLhogUJbvj99NPP8Uff/wBQRDw1ltvqZWJiHREtmTS0tJQUlKiVhbSKUmScOvWLSxevBiiKGodh1QmWzLt7e3o7e3lTIbumyRJaDxzHEulETSKqdi49yUWzTwjWzJbtmxBR0cH8vLy1MpDc4jZbEZfX9+sxoiOjsabuWnwiD4I0g0UFn4Ft9s962yZmZmoqqqa9TikPNmS6enpwRtvvAGbzYb169erlYnmiFDcxFMzmShpBH4xFd/8lzOZ+Ua2ZJxOJxoaGjA8PKxWHtIZURSxce9LGBkZwcbUVBbMPBTwKWyv1wu3242uri6YTCbExcUpHoZPYROFJ7l7N+BMprW1Fd9//z0A4OLFizh48KAy6YhI1wKWzFNPPQWHwwGfz4ehoSE1MxGRjsj+TMblcmF0dBTJyclq5SEinZF9rEAQBGzatAkGg0GtPESkM7Izmbt37yIzMxNLly5VKw8R6YzsTMbr9eL06dM4efKkWnmISGdkSyY1NRU+nw9paWlq5SEinZEtGY/HA6PRGJJt4EQ0P8n+TObOnTtoa2vDY489FvSA586dg91ux4IFC+Dz+TA+Po7XX38dUVFRsw5LROFHtmSioqJw+PBhLF68OOgBu7u7kZ6ejsnJSWzfvh0//vgjrly5wocsieYp2ZIpLS1FXV0dOjs7ceTIkaAGNJvNePDBB7FixQoUFRUF/Dy+TS3R/CD7DpIff/wxIiMj8fTTT8NkMgU1oM1mAwCMj49jfHwcgiDg1VdfRXR09D1fy2eXiMLTfT27BPz5/8kMDg5ieHg46JIpLy+fcUAi0i/Z1aWvv/4asbGxiI2NVSsPEenMPXf8ulwuCIKgVh4i0hnZmcymTZvgdDqxcOFCtfIQkc7IlkxLSwsmJibQ2tqqVh4i0pl7Plawe/dupKSkqJWHiHRGtmSMRiOOHj3KkiGi+yZbMjdv3sQrr7yC3t5etfIQkc7IlkxsbCzi4uKQmJioVh4i0hnZkjEYDPx2iYhmRXafTEFBAQoKCtTKQkQ6JDuTISKarXuWzJdffokPPvhAjSxEpEMBS+bs2bPo6upCfHw831qUiO5bwJJZt24dfvvtNzgcDpSVlamZiYh0JGDJ1NXVYWBgALt27UJbW5uamYhIRwKWzKJFi5CZmQlBELBlyxY1MxGRjgRcwn722WfVzEFEOsUlbCJSFEuGiBTFkiEiRbFkiAKQJAn9/f2QJEnrKGFNtyWj1wtEr+c110iShMYzxzFedwKNZ46H7d/3XLheZB+QnK2mpib8+uuvkCQJFotFyUP9w9QFslQaQaOYio17X9LFrmW9npdSzGYz+vr67uu10dHReDM3DR7RB0G6gcLCr0L6nvCZmZmoqqoK2Xj/Zq5cL4qWTFtbGyoqKnD06FF4vV4sWPDX4QK9g+Ty5ctnfdz4+Hgcys/GXXESk1IP1q49BZfLNetxtabX85qLIiIi0LQkEf9JjULLiBv9/f3w+XwhG394eDgk17qcOXO9+BX03nvvTf/q8XiUPFRQsrOztY4wK8yvrXDOr2V2RX8mk5OTA5vNhuTk5H/MYoho/lD0zn/mmWeUHJ6IwoBuV5eIaG6YVyVz8OBBrSPMCvNrK5zza5ld8Pv9fs2OTkS6N69mMkSkPpYMESlK1+vKkiTBarVizZo1WL16NU6ePAm32w2LxQKDwaB1vKCdO3cOdrsd6enp2Llzp9ZxZqy2thajo6OIiYnBc889p3WcGfn7NZSTk4PXXnsN1dXVWscK2lT+7OxsdHZ2QhRF7N+/H0ajUbUMup7JOBwOmEwmAEBrayvy8/NRXFyM5uZmbYPNUHd3N5KSksJ2r9HPP/+MF154AX19ffB6vVrHmZGpa8jv96O2thYrV67UOtKM/P0eGBsbw9jYGBYuXKhqBt2WTE1NDb799tvpxo6MjERERPidbk1NDRISElBeXo7r169jcnJS60gztnnzZpw4cQIOhyPs/g3S09NhNBoxNjaGyclJ/PTTT7h69arWsYI2lT8+Ph5msxn79u1DS0uLqhnC80tjEPbs2QMAuHTpEgAgLy8Pn332GURRhNls1jLajOzZswc2mw02mw0JCQmIjIzUOtKM2e12PPDAA3jiiSfCrmSmJCUlYceOHXA6nXj00Ue1jjNjQ0NDaG9vR0xMDEpLS1U9NpewiUhR4fllhYjCBkuGiBTFkiEiRbFkiEhRLBkKqQsXLqCqqgoulwsXLlyY8es///xzNDQ0TP/+wIEDYbVkTP9Pt0vYFFqHDh3CqlWr4HA4sGHDBtTV1cHj8aC0tBTV1dUwGAwoKytDR0cHbt26hRUrVqCjowNxcXG4evUqnE4nduzYgWPHjmHZsmXIy8vD6tWrMTExgY8++ghJSUnIysrC5cuXsX79egBAR0cHiouL0dDQEJbLxvQnzmQoKHFxcSgqKsLvv/+OpqYmmM1mlJSU4Nq1a1iyZAk8Hg8mJiYAAA899ND0ztjvvvsOBw4cwJNPPok7d+4gNzcXRUVF6O7uBgD09PRg1apVePnll3HlyhVkZWXh4YcfBgDU19fDbrejr68PAwMD2pw4zRpLhoIyNjaGM2fOICYmBhs2bEB1dTXOnz+PtWvXwuFwQBAE3L59GwBgMBhw8eJFAMC6detw/PhxtLe3IzExEYIgAACmtmdlZWXh2rVrOHXqFB5//PHp442OjiI5ORnPP/883n77bdTU1Kh8xhQq3IxHQTl8+DAqKiq0jkFhiCVDRIrit0tEpCiWDBEpiiVDRIpiyRCRolgyRKQolgwRKep/sMgRpc490WkAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 288x108 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "fig,ax = make_plot(nuc,[experiment])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "hide_input": false,
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  },
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}