{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Editing efficiency analysis\n",
    "\n",
    "This notebook contains the code to analyze the overall editing efficiency (by C position) for all the validation sgRNAs that we tested. This notebook takes as input the nucleotide_percentage files generated by the BEV_nucleotide_percentage_plots notebook."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "sys.path.append('../scripts/')\n",
    "import pandas as pd\n",
    "import matplotlib as mpl\n",
    "import seaborn as sns\n",
    "import notebook_functions as nb\n",
    "\n",
    "mpl.rc('pdf', fonttype=42)\n",
    "mpl.rcParams['font.sans-serif'] = \"Arial\"\n",
    "mpl.rcParams['font.family'] = \"sans-serif\"\n",
    "sns.set_context('paper')\n",
    "sns.set_style('ticks')\n",
    "boxprops = {'edgecolor': 'k', 'linewidth': 0.5, 'facecolor': 'w'}\n",
    "lineprops = {'color': 'k', 'linewidth': 0.5}\n",
    "stripplot_kwargs = dict({'linewidth': 0.5, 'size': 3, 'alpha': 0.8})\n",
    "boxplot_kwargs = dict({'boxprops': {'linewidth':0,'facecolor':'w'}, 'medianprops': {'linewidth':1},\n",
    "                       'whiskerprops': {'linewidth':0}, 'capprops': {'linewidth':0},\n",
    "                       'width': 0.8,'whis':(10,90)})\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## User inputs\n",
    "\n",
    "<font color='blue'> Please follow steps indicated in blue, then run the notebook to generate output files. If the files are formatted as described in the documentation, the code in the 'Functions' section should not need to be altered. </font> "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Metainformation file** \n",
    "\n",
    "<font color='blue'> <b>Step 1:</b> Please enter the filepath to the metainformation input file used in the allele frequencies notebook. The same file will be used in this notebook. </font> "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Please enter input filepath here: ../../AudreyData/TP53/Metainfo_input_ABE_TP53_fixed_updated_1d_sample.csv\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sg</th>\n",
       "      <th>sgRNA_sequence</th>\n",
       "      <th>translation_ref_seq</th>\n",
       "      <th>BEV_start</th>\n",
       "      <th>BEV_end</th>\n",
       "      <th>primer</th>\n",
       "      <th>frame</th>\n",
       "      <th>first_codon</th>\n",
       "      <th>last_codon</th>\n",
       "      <th>rev_com</th>\n",
       "      <th>BEV_ref</th>\n",
       "      <th>BEV_test</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1d</td>\n",
       "      <td>GCTCCTCCATGGCAGTGACC</td>\n",
       "      <td>[TTCCTCTTGCAGCAGCCAGACTGCCTTCCGGGTCACTGCC]ATGG...</td>\n",
       "      <td>417</td>\n",
       "      <td>426</td>\n",
       "      <td>F3_R2</td>\n",
       "      <td>1</td>\n",
       "      <td>ATG</td>\n",
       "      <td>CTG</td>\n",
       "      <td>True</td>\n",
       "      <td>417;418</td>\n",
       "      <td>425;426</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   sg        sgRNA_sequence  \\\n",
       "0  1d  GCTCCTCCATGGCAGTGACC   \n",
       "\n",
       "                                 translation_ref_seq  BEV_start  BEV_end  \\\n",
       "0  [TTCCTCTTGCAGCAGCCAGACTGCCTTCCGGGTCACTGCC]ATGG...        417      426   \n",
       "\n",
       "  primer  frame first_codon last_codon  rev_com  BEV_ref BEV_test  \n",
       "0  F3_R2      1         ATG        CTG     True  417;418  425;426  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "global old_input_file\n",
    "# input_filepath = 'AudreyData/Metainfo_input_f2_r1_ABE.csv'\n",
    "input_filepath = input(\"Please enter input filepath here: \")\n",
    "old_input_file = pd.read_csv(input_filepath)\n",
    "\n",
    "old_input_file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<font color='blue'><b>Step 2:</b> Enter filepath to folder where the files generated by the nucleotide percentage notebook were stored. Please make sure that the filepath ends in a '/'. </font> "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Please enter filepath to folder with nucleotide percentage outputs here: ../../AudreyData/TP53/ABE/nucleotide_percentage\n",
      "../../AudreyData/TP53/ABE/nucleotide_percentage/\n"
     ]
    }
   ],
   "source": [
    "# User inputs filepath where nucleotide_percentage outputs are stored\n",
    "global nuc_per_filepath\n",
    "# nuc_per_filepath = 'AudreyData/Validation_CRISPResso_results/nucleotide_percentage/'\n",
    "nuc_per_filepath = input(\"Please enter filepath to folder with nucleotide percentage outputs here: \")\n",
    "nuc_per_filepath = nb.check_folder_filepath(nuc_per_filepath)\n",
    "print(nuc_per_filepath)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<font color='blue'><b>Step 3:</b> Enter filepath to folder where the files generated by this notebook will be stored. Please make sure that the filepath ends in a '/'. If the folders in this file path do not currently exist, they will be created when the notebook is run.  </font> "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Please enter output folder filepath here: ../../AudreyData/TP53/ABE/\n",
      "../../AudreyData/TP53/ABE/\n"
     ]
    }
   ],
   "source": [
    "# User inputs filepath to store editing efficency outputs\n",
    "global output_filepath\n",
    "# output_filepath = 'AudreyData/Validation_CRISPResso_results/'\n",
    "output_filepath = input(\"Please enter output folder filepath here: \")\n",
    "output_filepath = nb.check_folder_filepath(output_filepath)\n",
    "print(output_filepath)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<font color='blue'><b>Step 4:</b> Enter the name of the gene targeted in experiment. </font> "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Please enter gene name here: TP53\n"
     ]
    }
   ],
   "source": [
    "# User inputs name of gene being targeted\n",
    "global experiment\n",
    "experiment = input(\"Please enter gene name here: \")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<font color='blue'><b>Step 5:</b> Please select the type of base editor (BE) used in the samples in input file. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Please specify the type of base editor used in the input samples by entering 'A' for A base editor or 'C' for C base editor: A\n"
     ]
    }
   ],
   "source": [
    "# User inputs type of base editor\n",
    "global be_type\n",
    "\n",
    "be_type_input = input(\"Please specify the type of base editor used in the input samples by entering 'A' for A base editor or 'C' for C base editor: \")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/javascript": [
       "IPython.notebook.execute_cells_below()"
      ],
      "text/plain": [
       "<IPython.core.display.Javascript object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Make sure a base editor is selected and not default value\n",
    "\n",
    "if (be_type_input != 'A') and (be_type_input != 'C'):\n",
    "    raise Exception('Invalid input. Please enter either A or C to specify base editor.')\n",
    "\n",
    "else:\n",
    "    be_type = be_type_input + 'BE'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "../../AudreyData/TP53/ABE/nucleotide_percentage/BEV_417_418_F3_R2.csv\n"
     ]
    }
   ],
   "source": [
    "# Generate editing efficiency table\n",
    "df = nb.run_editing_eff(old_input_file, nuc_per_filepath, output_filepath, experiment, be_type)\n",
    "save_filepath = output_filepath + 'editing_efficiency_input.csv'\n",
    "df.to_csv(save_filepath)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "df = df.copy().drop('sg', axis=1).drop_duplicates()\n",
    "if be_type == 'ABE':\n",
    "    df.loc[(df['Position'] >= 4) & (df['Position'] <= 8) & (df['Nucleotide'] == 'G'),'Position'].value_counts()\n",
    "elif be_type == 'CBE':\n",
    "    df.loc[(df['Position'] >= 4) & (df['Position'] <= 8) & (df['Nucleotide'] == 'T'),'Position'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Generate plots"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<font color='blue'><b>Step 6:</b> Enter the target nucleotide of interest (T, A, G, C, or del). You can re-run the following cells to change the nucleotide being plotted. </font> "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Please enter target nucleotide here: G\n"
     ]
    }
   ],
   "source": [
    "nuc = input(\"Please enter target nucleotide here: \")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAARkAAABlCAYAAAB5h5cZAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjQuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8rg+JYAAAACXBIWXMAAAsTAAALEwEAmpwYAAAKxElEQVR4nO3dXVBUZQMH8P8BOSAQH4u4SMNMjILrWNiQhpmvMeakiCMjkB/kuFhtZeNFU03clu9NWtZWNKWOugzOMDbjUEZcMA7QEOMI4ih9ATLStMiHbLhYK+zZZfe9aKCatz0usucc9vD/3Tgi+5z/0XP+PPA8xxX8fr8fREQKidA6ABHpG0uGiBTFkiEiRbFkiEhRLBkiUhRLhogUtUDrAKQvTqcT27Ztg9VqRUZGBm7fvg2TyTT9536/H1arFYODgxBFEW63GxUVFVi0aJGGqUlJAvfJUCgdO3YMycnJaG5uxpEjR1BTU4Pr169jzZo1KCwsRH19PSRJQllZGQDAbrfD5XL9o4hIXziToZDxeDxoaGjAF198gcbGRgwMDMBisaC3txfvvPMO3G43enp6YLFYAACVlZX45ZdfkJGRwZLRMc5kKGTOnz+P06dP45FHHsHg4CASEhIQERGBrKwslJSUICUlBWfPnsXExATMZjMAoL+/H5WVlXj33Xc1Tk9KYclQyOzduxcffvghUlNT4ff7sXv3brz//vvIyMiY/hyfzwer1YobN24gOjoaHo8HL774InJycjRMTkpiyRCRoriETUSKYskQkaJYMkSkqJAsYUuSBKvVitzcXPzwww9wu92wWCyor6+H1+uF0WhEQUFBKA5FRGEmJDMZh8MBk8kEl8uF/Px8FBcXo7m5GTdv3kR5eTk6OztDcRgiCkMhmcmkp6fDaDRicHAQERF/9VZkZGTA13zyySeorKwMxeGJaA7o7u7+14+HbAn70qVLGBoaQldXF0RRhNlsRm1tLURRREpKCrZu3XrPMZYvXx4wKBHNXXL37pzaJ8OSIQpPcvcuV5eISFEsGSJSFEuGiBTFkiEiRbFkiEhRLBkiUhRLhogUJbvj99NPP8Uff/wBQRDw1ltvqZWJiHREtmTS0tJQUlKiVhbSKUmScOvWLSxevBiiKGodh1QmWzLt7e3o7e3lTIbumyRJaDxzHEulETSKqdi49yUWzTwjWzJbtmxBR0cH8vLy1MpDc4jZbEZfX9+sxoiOjsabuWnwiD4I0g0UFn4Ft9s962yZmZmoqqqa9TikPNmS6enpwRtvvAGbzYb169erlYnmiFDcxFMzmShpBH4xFd/8lzOZ+Ua2ZJxOJxoaGjA8PKxWHtIZURSxce9LGBkZwcbUVBbMPBTwKWyv1wu3242uri6YTCbExcUpHoZPYROFJ7l7N+BMprW1Fd9//z0A4OLFizh48KAy6YhI1wKWzFNPPQWHwwGfz4ehoSE1MxGRjsj+TMblcmF0dBTJyclq5SEinZF9rEAQBGzatAkGg0GtPESkM7Izmbt37yIzMxNLly5VKw8R6YzsTMbr9eL06dM4efKkWnmISGdkSyY1NRU+nw9paWlq5SEinZEtGY/HA6PRGJJt4EQ0P8n+TObOnTtoa2vDY489FvSA586dg91ux4IFC+Dz+TA+Po7XX38dUVFRsw5LROFHtmSioqJw+PBhLF68OOgBu7u7kZ6ejsnJSWzfvh0//vgjrly5wocsieYp2ZIpLS1FXV0dOjs7ceTIkaAGNJvNePDBB7FixQoUFRUF/Dy+TS3R/CD7DpIff/wxIiMj8fTTT8NkMgU1oM1mAwCMj49jfHwcgiDg1VdfRXR09D1fy2eXiMLTfT27BPz5/8kMDg5ieHg46JIpLy+fcUAi0i/Z1aWvv/4asbGxiI2NVSsPEenMPXf8ulwuCIKgVh4i0hnZmcymTZvgdDqxcOFCtfIQkc7IlkxLSwsmJibQ2tqqVh4i0pl7Plawe/dupKSkqJWHiHRGtmSMRiOOHj3KkiGi+yZbMjdv3sQrr7yC3t5etfIQkc7IlkxsbCzi4uKQmJioVh4i0hnZkjEYDPx2iYhmRXafTEFBAQoKCtTKQkQ6JDuTISKarXuWzJdffokPPvhAjSxEpEMBS+bs2bPo6upCfHw831qUiO5bwJJZt24dfvvtNzgcDpSVlamZiYh0JGDJ1NXVYWBgALt27UJbW5uamYhIRwKWzKJFi5CZmQlBELBlyxY1MxGRjgRcwn722WfVzEFEOsUlbCJSFEuGiBTFkiEiRbFkiAKQJAn9/f2QJEnrKGFNtyWj1wtEr+c110iShMYzxzFedwKNZ46H7d/3XLheZB+QnK2mpib8+uuvkCQJFotFyUP9w9QFslQaQaOYio17X9LFrmW9npdSzGYz+vr67uu10dHReDM3DR7RB0G6gcLCr0L6nvCZmZmoqqoK2Xj/Zq5cL4qWTFtbGyoqKnD06FF4vV4sWPDX4QK9g+Ty5ctnfdz4+Hgcys/GXXESk1IP1q49BZfLNetxtabX85qLIiIi0LQkEf9JjULLiBv9/f3w+XwhG394eDgk17qcOXO9+BX03nvvTf/q8XiUPFRQsrOztY4wK8yvrXDOr2V2RX8mk5OTA5vNhuTk5H/MYoho/lD0zn/mmWeUHJ6IwoBuV5eIaG6YVyVz8OBBrSPMCvNrK5zza5ld8Pv9fs2OTkS6N69mMkSkPpYMESlK1+vKkiTBarVizZo1WL16NU6ePAm32w2LxQKDwaB1vKCdO3cOdrsd6enp2Llzp9ZxZqy2thajo6OIiYnBc889p3WcGfn7NZSTk4PXXnsN1dXVWscK2lT+7OxsdHZ2QhRF7N+/H0ajUbUMup7JOBwOmEwmAEBrayvy8/NRXFyM5uZmbYPNUHd3N5KSksJ2r9HPP/+MF154AX19ffB6vVrHmZGpa8jv96O2thYrV67UOtKM/P0eGBsbw9jYGBYuXKhqBt2WTE1NDb799tvpxo6MjERERPidbk1NDRISElBeXo7r169jcnJS60gztnnzZpw4cQIOhyPs/g3S09NhNBoxNjaGyclJ/PTTT7h69arWsYI2lT8+Ph5msxn79u1DS0uLqhnC80tjEPbs2QMAuHTpEgAgLy8Pn332GURRhNls1jLajOzZswc2mw02mw0JCQmIjIzUOtKM2e12PPDAA3jiiSfCrmSmJCUlYceOHXA6nXj00Ue1jjNjQ0NDaG9vR0xMDEpLS1U9NpewiUhR4fllhYjCBkuGiBTFkiEiRbFkiEhRLBkKqQsXLqCqqgoulwsXLlyY8es///xzNDQ0TP/+wIEDYbVkTP9Pt0vYFFqHDh3CqlWr4HA4sGHDBtTV1cHj8aC0tBTV1dUwGAwoKytDR0cHbt26hRUrVqCjowNxcXG4evUqnE4nduzYgWPHjmHZsmXIy8vD6tWrMTExgY8++ghJSUnIysrC5cuXsX79egBAR0cHiouL0dDQEJbLxvQnzmQoKHFxcSgqKsLvv/+OpqYmmM1mlJSU4Nq1a1iyZAk8Hg8mJiYAAA899ND0ztjvvvsOBw4cwJNPPok7d+4gNzcXRUVF6O7uBgD09PRg1apVePnll3HlyhVkZWXh4YcfBgDU19fDbrejr68PAwMD2pw4zRpLhoIyNjaGM2fOICYmBhs2bEB1dTXOnz+PtWvXwuFwQBAE3L59GwBgMBhw8eJFAMC6detw/PhxtLe3IzExEYIgAACmtmdlZWXh2rVrOHXqFB5//PHp442OjiI5ORnPP/883n77bdTU1Kh8xhQq3IxHQTl8+DAqKiq0jkFhiCVDRIrit0tEpCiWDBEpiiVDRIpiyRCRolgyRKQolgwRKep/sMgRpc490WkAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 288x108 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "fig,ax = nb.make_editing_eff_plot(df, nuc,[experiment], be_type, output_filepath)\n",
    "\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "hide_input": false,
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  },
  "varInspector": {
   "cols": {
    "lenName": 16,
    "lenType": 16,
    "lenVar": 40
   },
   "kernels_config": {
    "python": {
     "delete_cmd_postfix": "",
     "delete_cmd_prefix": "del ",
     "library": "var_list.py",
     "varRefreshCmd": "print(var_dic_list())"
    },
    "r": {
     "delete_cmd_postfix": ") ",
     "delete_cmd_prefix": "rm(",
     "library": "var_list.r",
     "varRefreshCmd": "cat(var_dic_list()) "
    }
   },
   "types_to_exclude": [
    "module",
    "function",
    "builtin_function_or_method",
    "instance",
    "_Feature"
   ],
   "window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}